src/base-subscription.ts

```diff
@@ -0,0 +1,253 @@
+                } else {
+                    // We consumed some response text, and all's fine. We expect more text.
+                }
+            }
+        }
+        
+        private onDone(): void {
+            if (this.xhr.status === 200) {
+                if (this.state === BaseSubscriptionState.OPENING) {
+                    this.state = BaseSubscriptionState.OPEN;
+                    this.onOpen(responseToHeadersObject(this.xhr.getAllResponseHeaders()));
+                }
+                this.assertStateIsIn( BaseSubscriptionState.OPEN, BaseSubscriptionState.ENDING );
+                let err = this.onChunk();
+                if (err) {
+                    this.state = BaseSubscriptionState.ENDED;
+                    if ( (err as any).statusCode === 204 ) { //TODO: That cast is horrific
+                        this.onEnd();
+                    }
+                    else {
+                        this.onError(err);
+                    }
+                } else if (this.state <= BaseSubscriptionState.ENDING) {
+                    this.onError(new Error("HTTP response ended without receiving EOS message"));
+                } else {
+                    // Stream ended normally.
+                    this.onEnd();
+                }
+            }
+            
+            else {
+                this.assertStateIsIn(BaseSubscriptionState.OPENING, BaseSubscriptionState.OPEN, BaseSubscriptionState.ENDED);
+                
+                if (this.state === BaseSubscriptionState.ENDED) {
+                    // We aborted the request deliberately, and called onError/onEnd elsewhere.
+                    return;
+                }
+                //Something terrible has happened. Most likely a network error. XHR is useless at that point.
+                else if(this.xhr.status === 0) {
+                    this.onError(new NetworkError("Connection lost."));
+                    
+                } else {
+                    this.onError(ErrorResponse.fromXHR(this.xhr));
+                }
+            }
+        }
+        
+        private lastNewlineIndex: number = 0;
+        
+        private onChunk(): Error {
+            this.assertStateIsIn(BaseSubscriptionState.OPEN);
+            let response = this.xhr.responseText;
+            let newlineIndex = response.lastIndexOf("\n");
+            if (newlineIndex > this.lastNewlineIndex) {
+                
+                let rawEvents = response.slice(this.lastNewlineIndex, newlineIndex).split("\n");
+                this.lastNewlineIndex = newlineIndex;
+                
+                for (let rawEvent of rawEvents) {
+                    if (rawEvent.length === 0) {
+                        continue; // FIXME why? This should be a protocol error
+                    }
+                    let data = JSON.parse(rawEvent);
+                    let err = this.onMessage(data);
+                    if (err != null) {
+                        return err;
```

Comment:
What happens if you have say 3 messages in the `rawEvents` variable but hit an error parsing the first 1? Are the other 2 lost? That's how it reads to me, but I might be missing something. Or it could be the desired behaviour?

Reply:
I didn't really touch that code, just moved it around. 
I believe this would be a caused network error, or some weird glitch in the application sending gibberish (not by protocol). 
I don't think we should try and do anything apart from retrying to subscribe in that case - if it's one-of the resumable sub _should_ handle it.  @pl any opinion?